src/infra/core/src/dependency_graph_service_inmem.rs

```diff
@@ -0,0 +1,316 @@
+};
+use kamu_core::*;
+use opendatafabric::DatasetID;
+use petgraph::stable_graph::{NodeIndex, StableDiGraph};
+use petgraph::Direction;
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+pub struct DependencyGraphServiceInMemory {
+    repository: Option<Arc<dyn DependencyGraphRepository>>,
+    state: Arc<tokio::sync::Mutex<State>>,
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Default)]
+struct State {
+    datasets_graph: StableDiGraph<DatasetID, ()>,
+    dataset_node_indices: HashMap<DatasetID, NodeIndex>,
+    initially_scanned: bool,
+}
+
+impl State {
+    fn get_dataset_node(
+        &self,
+        dataset_id: &DatasetID,
+    ) -> Result<NodeIndex, DatasetNodeNotFoundError> {
+        match self.dataset_node_indices.get(dataset_id) {
+            Some(index) => Ok(index.to_owned()),
+            None => Err(DatasetNodeNotFoundError {
+                dataset_id: dataset_id.clone(),
+            }),
+        }
+    }
+
+    fn get_or_create_dataset_node(&mut self, dataset_id: &DatasetID) -> NodeIndex {
+        match self.dataset_node_indices.get(dataset_id) {
+            Some(index) => index.to_owned(),
+            None => {
+                let node_index = self.datasets_graph.add_node(dataset_id.clone());
+                self.dataset_node_indices
+                    .insert(dataset_id.clone(), node_index);
+                node_index
+            }
+        }
+    }
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+#[component(pub)]
+#[interface(dyn DependencyGraphService)]
+#[interface(dyn AsyncEventHandler<DatasetEventCreated>)]
+#[interface(dyn AsyncEventHandler<DatasetEventDeleted>)]
+#[interface(dyn AsyncEventHandler<DatasetEventDependenciesUpdated>)]
+#[scope(Singleton)]
+impl DependencyGraphServiceInMemory {
+    pub fn new(repository: Option<Arc<dyn DependencyGraphRepository>>) -> Self {
+        Self {
+            repository,
+            state: Arc::new(tokio::sync::Mutex::new(State::default())),
+        }
+    }
+
+    async fn ensure_datasets_initially_scanned(&self) -> Result<(), InternalError> {
+        let mut state = self.state.lock().await;
+        if state.initially_scanned {
+            return Ok(());
+        }
+
+        self.ensure_datasets_initially_scanned_with(
+            &mut state,
+            self.repository
+                .as_ref()
+                .expect("Dependencies graph repository not present")
+                .as_ref(),
+        )
+        .await
+    }
+
+    async fn ensure_datasets_initially_scanned_with(
+        &self,
+        state: &mut State,
+        repository: &dyn DependencyGraphRepository,
+    ) -> Result<(), InternalError> {
+        if state.initially_scanned {
+            return Ok(());
+        }
+
+        tracing::debug!("Started initializing dependencies graph lazily");
```

Comment:
Instead of logging started / finished events consider separating initialization into its own function with a `#[tracing::instrument()]` macro. It will log the end event even if function terminates with error and also measure the wall clock time for us to know how long this span took.

Reply:
Done. Although, I've left a final trace with different text, which displays node/edge count in debug mode.